test/rule_creation_protocol/protocol.py

```diff
@@ -0,0 +1,301 @@
+        """ The rule code is made of 3 digits capital letters (taken from the list of Functional parts)
+        + 3 digit numbers 
+        e.g. 'ALB123', 'GEM391' and not 'AB14' or 'GKE9001'
+        """
+        validator_helper = ValidatorHelper()
+        validation = validator_helper.validate_rule_code(value)
+        if validation:
+            raise ProtocolError(
+                value=validation['value'],
+                message=validation['message']
+            )
+        return value
+
+    @field_validator('first_separator')
+    def validate_first_separator(cls, value, values):
+        valid_first_separator = values.data['valid_first_separator']
+        if value != valid_first_separator:
+            raise ProtocolError(
+                value=value,
+                message=f"expected separator {valid_first_separator} but instead found {value}"
+            )
+        return value, values
+
+    @field_validator('separators')
+    def validate_separators(cls, value, values):
+        separators = value
+        valid_separators = values.data['valid_separators']
+        unvalid_separators = [separator for separator in separators if separator != valid_separators]
+        if any(unvalid_separators):
+            raise ProtocolError(
+                value = separators,
+                message = f"expected {valid_separators} but found the following unvalid seperators {unvalid_separators}"
+            )
+        return value, values
+
+
+class RuleCreationConventions(ConfiguredBaseModel):
+    """Validate feature conventions
+    """
+    feature: Naming  # feature as in .feature file
+    dotfeature_file: Naming  # either user input or test file
+    ifc_input: dict
+    tags: list
+    description: str
+    steps: list
+    filename: str
+    readme: str
+
+    @field_validator('tags')
+    def do_validate_tags(cls, value) -> dict:
+        validator = ValidatorHelper()
+        validated_tags = validator.validate_tags(value)
+        if validated_tags != 'passed':
+            raise ProtocolError(
+                value=validated_tags['value'],
+                message=validated_tags['message']
+            )
+        return value
+
+    @field_validator('ifc_input')
+    def validate_ifc_input(cls, value):
+        # @todo implement
+        pass
+
+    @field_validator('description')
+    def validate_description(cls, value=list) -> list:
+        """must include a description of the rule that start with "The rule verifies that..."""  # allow for comma's
+        if not any(value.startswith(f"{prefix} rule verifies{optional_comma} that") for prefix in ("This", "The") for optional_comma in ("", ",")):
+            raise ProtocolError(
+                value=value,
+                message=f"The description must start with 'The rule verifies that', it now starts with {value}"
+            )
+        return value
+
+    @field_validator('steps')
+    def validate_steps(cls, value):
```

Comment:
This set of validations is checked for every test file within a rule. I think it would be enough to check it once per rule (feature), not once per test file per rule.

While this can be changed, I think it's not terrible to leave it as is, because our test_main.py procedure is based on looping through every test file. @Ghesselink , what do you think?

Reply:
That would be enough for all the conventions that are tested in the `.feature` file. 
However, there are also conventions related to each test file (the naming for instance, and having .ifc as an extention). 

The place in test_main.py seemed to be the most natural. The costs is that each .feature file is tested multiple times (with each test file), but it's quite cheap since it runs very quick.
An alternative is check whether a `.feature `file is already checked and register that somewhere so it passes that validation for the next run (with the next testfile). But maybe we shouldn't complicate it too much, it's already quite challenging to keep a good overview of all the conventions already.